src/petals/client/inference_session.py

```diff
@@ -94,6 +95,12 @@ def step(
         if self.closed:
             raise Exception("Session is closed, cannot perform step")
 
+        if last_validated_position is not None:
+            assert last_validated_position <= self._position
+            self._position = last_validated_position
+            if self.history is not None and self.history.shape[1] >= last_validated_position:
+                self.history = self.history[:, :last_validated_position, :] if last_validated_position > 0 else None
```

Comment:
Another rationale: technically, `last_validated_position` is not the index (0-based) of last validated position - it is higher by 1. Otherwise, history[:, :last_validated_position] would delete the last validated token from history. If i got it wrong, please explain.